src/java/org/apache/cassandra/db/streaming/ComponentManifest.java

```diff
@@ -0,0 +1,129 @@
+    private final long totalSize;
+
+    public ComponentManifest(Map<Component.Type, Long> componentManifest)
+    {
+        this.manifest = new LinkedHashMap<>(componentManifest);
+
+        long size = 0;
+        for (Map.Entry<Component.Type, Long> entry : this.manifest.entrySet())
+        {
+            size += entry.getValue();
+            this.components.add(Component.parse(entry.getKey().repr));
+        }
+
+        this.totalSize = size;
+    }
+
+    public Long getSizeForType(Component.Type type)
+    {
+        return manifest.get(type);
+    }
+
+    public long getTotalSize()
+    {
+        return totalSize;
+    }
+
+    public Set<Component> getComponents()
+    {
+        return Collections.unmodifiableSet(components);
+    }
+
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ComponentManifest that = (ComponentManifest) o;
+        return totalSize == that.totalSize &&
+               Objects.equals(manifest, that.manifest);
+    }
+
+    public int hashCode()
+    {
+
+        return Objects.hash(manifest, totalSize);
+    }
+
+    public static final IVersionedSerializer<ComponentManifest> serializer = new IVersionedSerializer<ComponentManifest>()
+    {
+        public void serialize(ComponentManifest manifest, DataOutputPlus out, int version) throws IOException
+        {
+            out.writeInt(manifest.manifest.size());
+            for (Map.Entry<Component.Type, Long> entry : manifest.manifest.entrySet())
+                serialize(entry.getKey(), entry.getValue(), out);
+        }
+
+        public ComponentManifest deserialize(DataInputPlus in, int version) throws IOException
+        {
+            LinkedHashMap<Component.Type, Long> components = new LinkedHashMap<>(Component.Type.values().length);
+
+            int size = in.readInt();
+            assert size >= 0 : "Invalid number of components";
+
+            for (int i = 0; i < size; i++)
+            {
+                Component.Type type = Component.Type.fromRepresentation(in.readByte());
+                long length = in.readLong();
+                components.put(type, length);
+            }
+
+            return new ComponentManifest(components);
+        }
+
+        public long serializedSize(ComponentManifest manifest, int version)
+        {
+            long size = 0;
+            size += TypeSizes.sizeof(manifest.manifest.size());
+            for (Map.Entry<Component.Type, Long> entry : manifest.manifest.entrySet())
+            {
+                size += TypeSizes.sizeof(entry.getKey().id);
+                size += TypeSizes.sizeof(entry.getValue());
+            }
+            return size;
+        }
+
+        private void serialize(Component.Type type, long size, DataOutputPlus out) throws IOException
+        {
+            out.writeByte(type.id);
```

Comment:
Encoding types as a single byte reduces the size of the message a little, but it has a cost of unnecessarily increasing complexity. We do this in `Verb`, but there it is for a good reason: we ser/deser `Verb`s *a lot*. Here we'd be better of with just relying on `valueOf()`, IMO.

Reply:
@jasobrown and I debated this. Basically, we chose to go with the byte approach as it seemed to be better than going with ordinals. In degenerate cases where we have a large number of SSTables, this will make a difference as we will be sending a lot of these messages around. unless you strongly oppose removing this, I am inclined to leaving this as-is.